File: others/python-sdk/src/autumn_sdk/_hooks/fail_open_hook.py
import json
import sys
from typing import Optional, Tuple, Union

import httpx

from autumn_sdk.sdkconfiguration import SDKConfiguration

from .types import (
    AfterErrorContext,
    AfterErrorHook,
    SDKInitHook,
)

FAIL_OPEN_OPERATION_IDS = {"check", "track", "getOrCreateCustomer"}

FAIL_OPEN_LOG_MESSAGE = (
    "[Autumn] Request failed \u2014 failing open. "
    "Learn more: https://docs.useautumn.com/documentation/fail-open"
)

FAIL_OPEN_BODIES: dict = {
    "check": {
        "allowed": True,
        "customer_id": None,
        "balance": None,
        "flag": None,
    },
    "track": {
        "customer_id": None,
        "value": 0,
        "balance": None,
    },
    "getOrCreateCustomer": {
        "id": None,
        "name": None,
        "email": None,
        "created_at": 0,
        "fingerprint": None,
        "stripe_id": None,
        "env": "live",
        "metadata": {},
        "send_email_receipts": False,
        "billing_controls": {},
        "subscriptions": [],
        "purchases": [],
        "balances": {},
        "flags": {},
    },
}


def _make_synthetic_response(status_code: int, body: Optional[dict] = None) -> httpx.Response:
    content = json.dumps(body).encode() if body else b""
    headers = {"content-type": "application/json"} if body else {}
    return httpx.Response(
        status_code=status_code,
        headers=headers,
        content=content,
    )


class _SafeSyncClient:
    """Wraps an httpx sync client to catch connection/timeout errors and return a 503."""

    def __init__(self, inner: httpx.Client):
        self._inner = inner

    def send(self, request: httpx.Request, **kwargs) -> httpx.Response:
        try:
            return self._inner.send(request, **kwargs)
        except (httpx.ConnectError, httpx.TimeoutException):
            return _make_synthetic_response(503)

    def build_request(self, *args, **kwargs) -> httpx.Request:
        return self._inner.build_request(*args, **kwargs)

    def close(self) -> None:
        self._inner.close()

    def __getattr__(self, name):
        return getattr(self._inner, name)


class _SafeAsyncClient:
    """Wraps an httpx async client to catch connection/timeout errors and return a 503."""

    def __init__(self, inner: httpx.AsyncClient):
        self._inner = inner

    async def send(self, request: httpx.Request, **kwargs) -> httpx.Response:
        try:
            return await self._inner.send(request, **kwargs)
        except (httpx.ConnectError, httpx.TimeoutException):
            return _make_synthetic_response(503)

    def build_request(self, *args, **kwargs) -> httpx.Request:
        return self._inner.build_request(*args, **kwargs)

    async def aclose(self) -> None:
        await self._inner.aclose()

    def __getattr__(self, name):
        return getattr(self._inner, name)


class FailOpenHook(SDKInitHook, AfterErrorHook):
    def __init__(self):
        self._enabled = True

    def sdk_init(self, config: SDKConfiguration) -> SDKConfiguration:
        if config.globals.fail_open is False:
            self._enabled = False
            return config

        self._enabled = True

        if config.client is not None and not isinstance(config.client, _SafeSyncClient):
            config.client = _SafeSyncClient(config.client)

        if config.async_client is not None and not isinstance(config.async_client, _SafeAsyncClient):
            config.async_client = _SafeAsyncClient(config.async_client)

        return config

    def after_error(
        self,
        hook_ctx: AfterErrorContext,
        response: Optional[httpx.Response],
        error: Optional[Exception],
    ) -> Union[Tuple[Optional[httpx.Response], Optional[Exception]], Exception]:
        if not self._enabled:
            return response, error

        if response is None or response.status_code < 500:
            return response, error

        if hook_ctx.operation_id not in FAIL_OPEN_OPERATION_IDS:
            return response, error

        body = FAIL_OPEN_BODIES.get(hook_ctx.operation_id)
        if body is None:
            return response, error

        print(FAIL_OPEN_LOG_MESSAGE, file=sys.stderr)
        print(
            f"  Operation: {hook_ctx.operation_id} | "
            f"Status: {response.status_code} | "
            f"Error: {error or 'Server error'}",
            file=sys.stderr,
        )

        return _make_synthetic_response(200, body), None

File: packages/sdk/src/hooks/failOpenHook.ts
import type { SDKOptions } from "../lib/config.js";
import { HTTPClient } from "../lib/http.js";
import type {
	AfterErrorContext,
	AfterErrorHook,
	Awaitable,
	SDKInitHook,
} from "./types.js";

const FAIL_OPEN_OPERATION_IDS = new Set(["check", "track"]);

const FAIL_OPEN_LOG_MESSAGE =
	"[Autumn] Request failed — failing open. Learn more: https://docs.useautumn.com/documentation/fail-open";

const FAIL_OPEN_BODIES: Record<string, object> = {
	check: {
		allowed: true,
		customer_id: null,
		balance: null,
		flag: null,
	},
	track: {
		customer_id: null,
		value: 0,
		balance: null,
	},
};

export class FailOpenHook implements SDKInitHook, AfterErrorHook {
	private enabled = true;

	sdkInit(opts: SDKOptions): SDKOptions {
		if (opts.failOpen === false) {
			this.enabled = false;
			return opts;
		}

		this.enabled = true;

		opts.httpClient = new HTTPClient({
			fetcher: async (input, init) => {
				try {
					return init == null ? await fetch(input) : await fetch(input, init);
				} catch (error) {
					console.log(error);
					console.log(
						`Network failed to reach Autumn: ${error}. Returning 555 Network Error.`,
					);
					return new Response(null, {
						status: 555,
						statusText: "Network Error",
					});
				}
			},
		});

		return opts;
	}

	afterError(
		hookCtx: AfterErrorContext,
		response: Response | null,
		error: unknown,
	): Awaitable<{ response: Response | null; error: unknown }> {
		if (!this.enabled) {
			return { response, error };
		}

		if (!response || response.status < 500) {
			return { response, error };
		}

		if (!FAIL_OPEN_OPERATION_IDS.has(hookCtx.operationID)) {
			return { response, error };
		}

		const body = FAIL_OPEN_BODIES[hookCtx.operationID];
		if (!body) {
			return { response, error };
		}

		console.error(FAIL_OPEN_LOG_MESSAGE);
		console.error(
			`  Operation: ${hookCtx.operationID} | Status: ${response.status} | Error: ${error ?? "Server error"}`,
		);

		return {
			response: new Response(JSON.stringify(body), {
				status: 200,
				headers: { "Content-Type": "application/json" },
			}),
			error: null,
		};
	}
}
